File: web/lib/airspace.ts
/**
 * Airspace membership — which controlled volume an aircraft occupies.
 *
 * Pure, dependency-free point-in-polygon (the project ships no turf) over the
 * Bangkok airspace GeoJSON already loaded by `fetchSector` (web/lib/geojson.ts):
 * BACC sectors + subsectors, Control Zones (CTR), Terminal Areas (TMA) and
 * Prohibited/Danger/Restricted areas (PDR). The test is ALTITUDE-AWARE — a plane
 * at FL196 is not "in" a CTR that tops at 2000 ft — so each layer's vertical band
 * is parsed (their formats differ: numeric FL on bacc, numeric ft on tma, and
 * strings like "GND"/"ALT 2000"/"FL 120"/"UNL" on ctr/pdr).
 */

import type { Geometry, Position } from "geojson";

import { SECTORS, type SectorCollection, type SectorKey } from "./geojson";

export interface AirspaceMembership {
  bacc?: string; // BACC sector, e.g. "4S"
  subsector?: string; // BACC subsector, e.g. "7N"
  ctr?: string; // e.g. "BANGKOK CTR"
  tma?: string; // e.g. "BANGKOK TMA"
  pdr?: string[]; // e.g. ["VTD16 RATCHABURI"] — can overlap, so a list
}

// --- point-in-polygon (ray casting, lon/lat) -------------------------------

function pointInRing(lon: number, lat: number, ring: Position[]): boolean {
  let inside = false;
  for (let i = 0, j = ring.length - 1; i < ring.length; j = i++) {
    const xi = ring[i][0];
    const yi = ring[i][1];
    const xj = ring[j][0];
    const yj = ring[j][1];
    const intersect =
      yi > lat !== yj > lat &&
      lon < ((xj - xi) * (lat - yi)) / (yj - yi) + xi;
    if (intersect) inside = !inside;
  }
  return inside;
}

/** A GeoJSON Polygon is [outer, ...holes]. Inside the outer ring, not a hole. */
function pointInPolygon(lon: number, lat: number, poly: Position[][]): boolean {
  if (poly.length === 0 || !pointInRing(lon, lat, poly[0])) return false;
  for (let h = 1; h < poly.length; h++) {
    if (pointInRing(lon, lat, poly[h])) return false;
  }
  return true;
}

export function pointInMultiPolygon(
  lon: number,
  lat: number,
  mp: Position[][][],
): boolean {
  for (const poly of mp) if (pointInPolygon(lon, lat, poly)) return true;
  return false;
}

// --- edge-bucketed ring index ------------------------------------------------
//
// The ray cast above walks EVERY edge of a ring for every point tested. The
// airspace polygons are not small — a BACC sector runs to 28 000 vertices and a
// CTR to ~1 000 — so a single aircraft sample cost ~10 000 edge tests, and a
// whole traffic day (hundreds of flights, thousands of samples each) spent
// seconds in this loop. A horizontal ray only ever crosses edges whose latitude
// span contains the point, so the edges are bucketed by latitude band once and
// a point scans just its own band.
//
// EXACT, not approximate: a band holds every edge whose span overlaps it (a
// superset of the edges that can straddle the point), and each edge that is
// visited is put through the very same test as the plain loop, so the parity —
// and therefore the answer — is identical. Only edges that could never have
// toggled it are skipped.

/** Rings shorter than this are scanned plainly: the lookup would cost more than
 *  the loop it saves. */
const RING_INDEX_MIN_VERTICES = 64;

interface RingIndex {
  minY: number;
  maxY: number;
  /** Bands per degree of latitude. */
  scale: number;
  /** Index of the last band. */
  last: number;
  /** Band b holds edges[start[b] .. start[b + 1]). */
  start: Int32Array;
  /** Each entry is the vertex i of an edge (ring[j] → ring[i], j = i − 1 wrapped). */
  edges: Int32Array;
}

function buildRingIndex(ring: Position[]): RingIndex {
  const n = ring.length;
  let minY = Infinity;
  let maxY = -Infinity;
  for (let i = 0; i < n; i++) {
    const y = ring[i][1];
    if (y < minY) minY = y;
    if (y > maxY) maxY = y;
  }
  const bands = Math.max(16, Math.min(512, n >> 4));
  const span = maxY - minY;
  const scale = span > 0 ? bands / span : 0;
  const last = bands - 1;
  const bandOf = (y: number): number => {
    const b = Math.floor((y - minY) * scale);
    return b < 0 ? 0 : b > last ? last : b;
  };
  const start = new Int32Array(bands + 1);
  // Two passes over the same edges: count per band, then fill.
  for (let i = 0, j = n - 1; i < n; j = i++) {
    const yi = ring[i][1];
    const yj = ring[j][1];
    if (yi === yj) continue; // a level edge can never straddle a latitude
    const lo = bandOf(yi < yj ? yi : yj);
    const hi = bandOf(yi < yj ? yj : yi);
    for (let b = lo; b <= hi; b++) start[b + 1]++;
  }
  for (let b = 0; b < bands; b++) start[b + 1] += start[b];
  const edges = new Int32Array(start[bands]);
  const fill = start.slice(0, bands);
  for (let i = 0, j = n - 1; i < n; j = i++) {
    const yi = ring[i][1];
    const yj = ring[j][1];
    if (yi === yj) continue;
    const lo = bandOf(yi < yj ? yi : yj);
    const hi = bandOf(yi < yj ? yj : yi);
    for (let b = lo; b <= hi; b++) edges[fill[b]++] = i;
  }
  return { minY, maxY, scale, last, start, edges };
}

/** Same answer as `pointInRing`, scanning only the point's latitude band. */
function pointInRingIndexed(
  lon: number,
  lat: number,
  ring: Position[],
  idx: RingIndex,
): boolean {
  // Above or below every vertex: no edge straddles the latitude.
  if (lat < idx.minY || lat > idx.maxY) return false;
  const n = ring.length;
  let b = Math.floor((lat - idx.minY) * idx.scale);
  b = b < 0 ? 0 : b > idx.last ? idx.last : b;
  let inside = false;
  for (let k = idx.start[b], end = idx.start[b + 1]; k < end; k++) {
    const i = idx.edges[k];
    const j = i === 0 ? n - 1 : i - 1;
    const xi = ring[i][0];
    const yi = ring[i][1];
    const xj = ring[j][0];
    const yj = ring[j][1];
    const intersect =
      yi > lat !== yj > lat &&
      lon < ((xj - xi) * (lat - yi)) / (yj - yi) + xi;
    if (intersect) inside = !inside;
  }
  return inside;
}

/** One index per ring of one feature's MultiPolygon, parallel to `mp`
 *  (`undefined` for a ring too short to be worth one). */
type MultiPolygonIndex = (RingIndex | undefined)[][];

function buildMultiPolygonIndex(mp: Position[][][]): MultiPolygonIndex {
  return mp.map((poly) =>
    poly.map((ring) =>
      ring.length >= RING_INDEX_MIN_VERTICES ? buildRingIndex(ring) : undefined,
    ),
  );
}

/** `pointInMultiPolygon`, using the prebuilt ring indexes. */
function pointInMultiPolygonIndexed(
  lon: number,
  lat: number,
  mp: Position[][][],
  acc: MultiPolygonIndex,
): boolean {
  for (let p = 0; p < mp.length; p++) {
    const poly = mp[p];
    if (poly.length === 0) continue;
    const rings = acc[p];
    const inRing = (r: number): boolean => {
      const ix = rings[r];
      return ix
        ? pointInRingIndexed(lon, lat, poly[r], ix)
        : pointInRing(lon, lat, poly[r]);
    };
    if (!inRing(0)) continue;
    let inHole = false;
    for (let h = 1; h < poly.length; h++) {
      if (inRing(h)) {
        inHole = true;
        break;
      }
    }
    if (!inHole) return true;
  }
  return false;
}

// --- vertical bands --------------------------------------------------------

/** Feet from a limit value. `isFL` treats a bare number as a flight level.
 *  Strings: GND/SFC/MSL -> 0, UNL -> Infinity, "FL 120" -> 12000,
 *  "ALT 2000"/bare digits -> that many feet. */
export function parseAltFt(v: unknown, isFL = false): number {
  if (v == null) return NaN;
  if (typeof v === "number") return isFL ? v * 100 : v;
  const s = String(v).trim().toUpperCase();
  if (!s) return NaN;
  // "SURFACE" is spelled out on some PDR features (VTD74) where the rest use
  // GND. It used to fall through to the digit search, return NaN, and be
  // rescued only by the caller's fallback — right by luck, not by reading.
  if (s === "GND" || s === "SFC" || s === "MSL" || s === "SURFACE") return 0;
  if (s.startsWith("UNL")) return Infinity;
  const fl = s.match(/FL\s*(\d+)/);
  if (fl) return Number(fl[1]) * 100;
  const n = s.match(/(\d+)/);
  return n ? Number(n[1]) : NaN;
}

export interface Band {
  lo: number;
  hi: number;
}

/** Vertical band (feet) coded on a sector feature, per its layer's schema.
 *  Exposed so the map can colour sectors by altitude. */
export function layerBand(props: Record<string, unknown>, key: SectorKey): Band {
  switch (key) {
    case "bacc":
      return {
        lo: parseAltFt(props.lower, true),
        hi: parseAltFt(props.upper, true),
      };
    case "subsector":
      return { lo: 0, hi: Infinity }; // no coded band — horizontal only
    case "tma":
      return { lo: parseAltFt(props.lower), hi: parseAltFt(props.upper) };
    case "ctr":
      return {
        lo: parseAltFt(props.lower_1 ?? props.lowerlimit),
        hi: parseAltFt(props.upper_1 ?? props.upperlimit),
      };
    case "pdr":
      return {
        lo: parseAltFt(props.lowerlimit),
        hi: parseAltFt(props.upperlimit),
      };
  }
}

function layerLabel(props: Record<string, unknown>, key: SectorKey): string {
  if (key === "pdr") {
    const ident = String(props.ident ?? "").trim();
    const name = String(props.name ?? "").trim();
    return [ident, name].filter(Boolean).join(" ") || "PDR";
  }
  return String(props.name ?? props.ident ?? "").trim() || key.toUpperCase();
}

// --- prebuilt index (bbox + normalized MultiPolygon per feature) -----------

/** One indexed airspace volume: its published name, vertical band, bbox and
 *  normalised geometry. Exported because callers that reason about the SHAPE of
 *  the airspace — sector adjacency, say — must read the same volumes the
 *  membership test uses, or they end up describing a different airspace. */
export interface IndexEntry {
  label: string;
  band: Band;
  bbox: [number, number, number, number]; // minLon, minLat, maxLon, maxLat
  mp: Position[][][];
  /** Edge-bucketed ring indexes for `mp` (see RingIndex). Built with the index
   *  and optional: an entry assembled by hand still works, just unaccelerated. */
  acc?: MultiPolygonIndex;
}

export type AirspaceIndex = Partial<Record<SectorKey, IndexEntry[]>>;

function featureMP(geom: Geometry | null | undefined): Position[][][] | null {
  if (!geom) return null;
  if (geom.type === "MultiPolygon") return geom.coordinates as Position[][][];
  if (geom.type === "Polygon") return [geom.coordinates as Position[][]];
  return null;
}

function bboxOf(mp: Position[][][]): [number, number, number, number] {
  let minLon = Infinity;
  let minLat = Infinity;
  let maxLon = -Infinity;
  let maxLat = -Infinity;
  for (const poly of mp)
    for (const ring of poly)
      for (const p of ring) {
        const lon = p[0];
        const lat = p[1];
        if (lon < minLon) minLon = lon;
        if (lon > maxLon) maxLon = lon;
        if (lat < minLat) minLat = lat;
        if (lat > maxLat) maxLat = lat;
      }
  return [minLon, minLat, maxLon, maxLat];
}

/** Build a reusable index from the loaded sector collections (bbox precomputed
 *  once so per-frame membership is a cheap bbox-reject then ray-cast). */
export function buildAirspaceIndex(
  sectorData: Partial<Record<SectorKey, SectorCollection | null>>,
): AirspaceIndex {
  const idx: AirspaceIndex = {};
  for (const key of Object.keys(sectorData) as SectorKey[]) {
    const fc = sectorData[key];
    if (!fc) continue;
    const entries: IndexEntry[] = [];
    for (const f of fc.features) {
      const mp = featureMP(f.geometry);
      if (!mp) continue;
      const props = (f.properties ?? {}) as Record<string, unknown>;
      entries.push({
        label: layerLabel(props, key),
        band: layerBand(props, key),
        bbox: bboxOf(mp),
        mp,
        acc: buildMultiPolygonIndex(mp),
      });
    }
    idx[key] = entries;
  }
  return idx;
}

// --- membership ------------------------------------------------------------

const LAYER_ORDER: SectorKey[] = ["bacc", "subsector", "ctr", "tma", "pdr"];

/** Which airspace volumes contain (lon, lat, altFt). `altFt == null` (no
 *  vertical profile) skips the altitude gate → horizontal-only. */
export function airspaceAt(
  index: AirspaceIndex,
  lon: number,
  lat: number,
  altFt: number | null,
): AirspaceMembership {
  const m: AirspaceMembership = {};
  for (const key of LAYER_ORDER) {
    const entries = index[key];
    if (!entries) continue;
    const hits: string[] = [];
    for (const e of entries) {
      if (lon < e.bbox[0] || lon > e.bbox[2] || lat < e.bbox[1] || lat > e.bbox[3])
        continue;
      if (
        !(e.acc
          ? pointInMultiPolygonIndexed(lon, lat, e.mp, e.acc)
          : pointInMultiPolygon(lon, lat, e.mp))
      )
        continue;
      if (altFt != null && key !== "subsector") {
        if (!(altFt >= e.band.lo && altFt <= e.band.hi)) continue;
      }
      hits.push(e.label);
      if (key !== "pdr") break; // a plane is in exactly one of these
    }
    if (hits.length === 0) continue;
    if (key === "pdr") m.pdr = hits;
    else if (key === "bacc") m.bacc = hits[0];
    else if (key === "subsector") m.subsector = hits[0];
    else if (key === "ctr") m.ctr = hits[0];
    else if (key === "tma") m.tma = hits[0];
  }
  return m;
}

// --- whole-route segments (for the altitude profile block colouring) -------

/** Per-layer sector colour (the same palette that styles the map overlays). */
const SECTOR_COLOR = Object.fromEntries(
  SECTORS.map((s) => [s.key, s.color]),
) as Record<SectorKey, string>;

/** The map colour of the ONE airspace that owns the aircraft here — the layer
 *  the hierarchy resolves to (see {@link controllingLayer}). Returned as an
 *  array so the profile's fill helper keeps one shape; [] when in no airspace,
 *  which draws the neutral default. */
export function membershipColors(m: AirspaceMembership | undefined): string[] {
  const key = controllingLayer(m);
  return key === null ? [] : [SECTOR_COLOR[key]];
}

/** One contiguous stretch of a route that stays inside the same set of
 *  airspace volumes. `t0`/`t1` are seconds from the route's first point (the
 *  same clock the altitude chart uses), so the profile can paint the run as a
 *  colour block. Boundaries are stitched so adjacent segments abut exactly. */
export interface AirspaceSegment {
  t0: number;
  t1: number;
  /** Altitude-aware membership — the volumes that actually contain the
   *  aircraft on this stretch. Drives both the label and the colours. */
  membership: AirspaceMembership;
  /** Compact display label, e.g. "8S/Bangkok CTR/VTR1" ("" outside all zones).
   *  Altitude-aware: a TMA the aircraft is above (past its ceiling) is not
   *  listed here. */
  label: string;
  /** Per-layer colours to blend for this block ([] outside all zones) — from
   *  the same altitude-aware membership. */
  colors: string[];
}

interface SegPoint {
  lon: number;
  lat: number;
  altitude_ft: number | null;
  epoch_ts: string;
}

// One route's segments depend only on its points and the sector index, so the
// result is cached against the point array. A ray-cast per point over every
// sector is cheap for ONE route and ruinous for a whole traffic day: replacing
// a single flight (an applied CD&R fix) changes the trajectory ARRAY, and
// without this cache every one of the other flights was re-walked point by
// point on each Apply.
const segmentCache = new WeakMap<
  object,
  { index: AirspaceIndex; segs: AirspaceSegment[] }
>();

/** Walk a whole trajectory and collapse it into contiguous airspace segments.
 *  Membership is altitude-aware (a climb out of a low CTR drops that zone from
 *  both the label and the tint), so a new block starts wherever the set of
 *  containing volumes changes. Memoised per point array (see `segmentCache`) —
 *  one walk per route, not one per render. */
export function buildAirspaceSegments(
  index: AirspaceIndex,
  points: ReadonlyArray<SegPoint>,
): AirspaceSegment[] {
  if (points.length === 0 || !index.bacc) return [];
  // Cached against the index too: reloading the sector polygons builds a new
  // index, and the old segments were resolved against the old volumes.
  const hit = segmentCache.get(points);
  if (hit && hit.index === index) return hit.segs;
  const base = new Date(points[0].epoch_ts).getTime();
  let cur: AirspaceSegment | null = null;
  const segs: AirspaceSegment[] = [];
  for (const p of points) {
    const t = (new Date(p.epoch_ts).getTime() - base) / 1000;
    const full = airspaceAt(index, p.lon, p.lat, p.altitude_ft ?? null);
    // PDR (prohibited/danger/restricted) is EXCLUDED from the profile blocks:
    // those areas are assumed CLOSED (a flight wouldn't be routed through an
    // active one), so they must not tint or label the altitude chart as a
    // "sector". Only the controlling ATS volume (CTR/TMA/BACC/subsector) counts.
    const m = full.pdr ? { ...full, pdr: undefined } : full;
    const label = formatAirspace(m, "compact");
    if (cur && cur.label === label) {
      cur.t1 = t;
    } else {
      if (cur) cur.t1 = t; // stitch: previous block runs up to this transition
      cur = { t0: t, t1: t, membership: m, label, colors: membershipColors(m) };
      segs.push(cur);
    }
  }
  if (segs.length) segs[0].t0 = 0; // first block anchors at the departure edge
  segmentCache.set(points, { index, segs });
  return segs;
}

export function isEmptyAirspace(m: AirspaceMembership | undefined): boolean {
  return (
    !m ||
    (!m.bacc && !m.subsector && !m.ctr && !m.tma && (!m.pdr || m.pdr.length === 0))
  );
}

// Abbreviations that stay upper-case in a title-cased zone name.
const ZONE_ABBR = new Set([
  "CTR",
  "TMA",
  "FIR",
  "ACC",
  "CTA",
  "ATZ",
  "TCA",
  "MTMA",
  "APP",
]);

/** Title-case a zone name for display, keeping airspace abbreviations
 *  ("CTR"/"TMA") and area idents (anything with a digit, e.g. "VTD16")
 *  as-is: "BANGKOK TMA" → "Bangkok TMA", "VTD16 RATCHABURI" → "VTD16
 *  Ratchaburi", "NAN TMA" → "Nan TMA". */
function titleZone(name: string): string {
  return name
    .split(/\s+/)
    .map((w) => {
      if (/\d/.test(w)) return w;
      if (ZONE_ABBR.has(w.toUpperCase())) return w.toUpperCase();
      return w.charAt(0).toUpperCase() + w.slice(1).toLowerCase();
    })
    .join(" ");
}

/** Sector code for display. 3S and 6S are modelled as two altitude slabs
 *  (``3S_lower`` FL0–270, ``3S_upper`` FL270–460 — the part below 2S vs beside
 *  it), but a target is called just "3S" / "6S" regardless of slab (per BACC
 *  ops): strip the ``_lower``/``_upper`` suffix. The altitude test already put
 *  the aircraft in the right slab; the suffix is a data-modelling detail, not
 *  something a controller says. */
function sectorLabel(code: string): string {
  return code.replace(/_(lower|upper)$/i, "");
}

/**
 * The name a sector event carries for one indexed volume.
 *
 * `IndexEntry.label` is the raw published name; the events and the report rows
 * carry the DISPLAY name that `formatAirspace` produces. Anything joining the
 * two — the adjacency graph behind dynamic sectorization — has to cross that
 * gap with the same rules, or "3S_upper" and "3S" become two sectors that never
 * meet.
 */
export function sectorDisplayName(layer: SectorKey, label: string): string {
  if (layer === "pdr") return label;
  if (layer === "ctr" || layer === "tma") return titleZone(label);
  return sectorLabel(label);
}

/** Airspace hierarchy — an aircraft is in exactly ONE airspace at a time, so a
 *  point that falls inside several overlapping volumes resolves to one. Order
 *  (highest first), per the BACC ops structure:
 *
 *    1. **PDR** — prohibited/danger/restricted. Not an ATS unit, but being
 *       inside one is the fact that matters, so it overrides.
 *    2. **CTR** — Control Zone, worked by Aerodrome Control (Tower).
 *    3. **TMA** — Terminal Control Area, worked by Approach Control.
 *       (A CTA would sit here too — none in the Thai dataset.)
 *    4. **BACC sector** — Area Control (ACC). This is the reporting unit; it
 *       carries the vertical limits (2S FL270–460, 3S/6S below).
 *    5. **subsector** — the horizontal controller-split, used only when a
 *       sector is divided. It sits INSIDE its sector, so with the sector
 *       above it a target normally reports the sector; the subsector shows
 *       only where no sector is defined.
 *
 *  Annex 11 airspace/ATS-unit structure, resolved AFTER the lateral and
 *  vertical tests — an aircraft above a CTR's ceiling has already dropped out
 *  of it and falls through to the TMA/ACC below.
 *  Must stay in step with _HIERARCHY in trajectory_sim/airspace.py. */
const HIERARCHY = ["pdr", "ctr", "tma", "bacc", "subsector"] as const;

/** Which layer owns the aircraft here, or null when it is in none. */
export function controllingLayer(
  m: AirspaceMembership | undefined,
): SectorKey | null {
  if (isEmptyAirspace(m)) return null;
  const mm = m as AirspaceMembership;
  for (const key of HIERARCHY) {
    if (key === "pdr" ? mm.pdr?.length : mm[key]) return key;
  }
  return null;
}

/** Human string for the ONE airspace that owns the aircraft (see
 *  {@link HIERARCHY}). `compact` (plane label / graph) shows a PDR by its
 *  ident, `full` (Results rows) spells it out. "" when in no airspace. */
export function formatAirspace(
  m: AirspaceMembership | undefined,
  mode: "compact" | "full",
): string {
  const key = controllingLayer(m);
  if (key === null) return "";
  const mm = m as AirspaceMembership;
  switch (key) {
    case "pdr":
      return mode === "compact"
        ? (mm.pdr as string[]).map((p) => p.split(" ")[0]).join(",")
        : (mm.pdr as string[]).map(titleZone).join(" · ");
    case "ctr":
      return titleZone(mm.ctr as string);
    case "tma":
      return titleZone(mm.tma as string);
    default:
      return sectorLabel(mm[key] as string);
  }
}

File: trajectory_sim/airspace.py
"""Airspace sector membership for exports — Python twin of web/lib/airspace.ts.

Tags each trajectory sample with the airspace volumes that contain it, so the
download formats (CSV / GeoPackage / GeoJSON) can carry a per-timestamp
"sector" column that matches what the web UI shows on the map and the
altitude-profile chart. Reads the SAME GeoJSON sector files the web app
renders (web/public/data/sectors/): BACC sectors + subsectors, Control Zones
(CTR), Terminal Areas (TMA) and Prohibited/Danger/Restricted areas (PDR).

Membership is ALTITUDE-AWARE: a point belongs only to volumes whose vertical
band contains its altitude ("which sector is the aircraft IN") — a plane at
FL350 is not in a TMA that tops at 11 000 ft. The corrected vertical limits
(sectors_corrected/, fixed against AIP Thailand) are therefore what decides
each label.

The vertical-limit formats differ per layer exactly as on the web: numeric FL
on bacc, numeric feet on tma, and strings like "GND"/"ALT 2000"/"FL 120"/
"UNL" on ctr/pdr. Labels are rendered with the web's compact format, e.g.
``"8S/Bangkok CTR/VTR1"``, and layer hits follow the web's rule: first
feature in file order wins (PDR areas overlap, so they are all listed).
"""

from __future__ import annotations

import json
import math
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Sequence

import numpy as np
import shapely
from shapely.geometry import shape

# Sector GeoJSONs live with the web app's static data so both sides of the
# stack read the one dataset. `sectors_corrected` has the vertical limits fixed
# against AIP Thailand ENR 2.1 / 5.1 (AIRAC 2026-07-09) — see its CORRECTIONS.md.
# Layer order = display order in the compact label.
_SECTORS_DIR = (
    Path(__file__).resolve().parents[1]
    / "web"
    / "public"
    / "data"
    / "sectors_corrected"
)
_LAYERS: list[tuple[str, str]] = [
    ("bacc", "bacc_geo"),
    ("subsector", "bacc_subsector"),
    ("ctr", "ctr"),
    ("tma", "tma"),
    ("pdr", "pdr"),
]

# Abbreviations kept upper-case when title-casing a zone name for display.
_ZONE_ABBR = {"CTR", "TMA", "FIR", "ACC", "CTA", "ATZ", "TCA", "MTMA", "APP"}

#: Airspace hierarchy — an aircraft is in exactly ONE airspace at a time, so a
#: point inside several overlapping volumes resolves to one (highest priority
#: first), per the BACC ops structure:
#:
#:   1. ``pdr``       — prohibited/danger/restricted. Not an ATS unit, but being
#:                      inside one is the fact that matters, so it overrides.
#:   2. ``ctr``       — Control Zone, worked by Aerodrome Control (Tower).
#:   3. ``tma``       — Terminal Control Area, worked by Approach Control.
#:                      (A CTA would sit here too — none in the Thai dataset.)
#:   4. ``bacc``      — Area Control (ACC). The reporting unit; it carries the
#:                      vertical limits (2S FL270-460, 3S/6S below).
#:   5. ``subsector`` — the horizontal controller-split, INSIDE its sector, so
#:                      with the sector above it a target normally reports the
#:                      sector; the subsector shows only where no sector exists.
#:
#: Annex 11 airspace / ATS-unit structure, applied AFTER the lateral and
#: vertical tests, so an aircraft above a CTR's ceiling has already dropped out
#: of it and falls through to the TMA/ACC below.
#: Must stay in step with HIERARCHY in web/lib/airspace.ts.
_HIERARCHY = ("pdr", "ctr", "tma", "bacc", "subsector")


def parse_alt_ft(v: object, is_fl: bool = False) -> float:
    """Feet from a vertical-limit value (mirror of the web's parseAltFt).

    ``is_fl`` treats a bare number as a flight level. Strings: GND/SFC/MSL ->
    0, UNL -> +inf, "FL 120" -> 12000, "ALT 2000"/bare digits -> feet.
    """
    if v is None:
        return math.nan
    if isinstance(v, (int, float)):
        if isinstance(v, float) and math.isnan(v):
            return math.nan
        return float(v) * 100.0 if is_fl else float(v)
    s = str(v).strip().upper()
    if not s:
        return math.nan
    if s in ("GND", "SFC", "MSL"):
        return 0.0
    if s.startswith("UNL"):
        return math.inf
    fl = re.search(r"FL\s*(\d+)", s)
    if fl:
        return float(fl.group(1)) * 100.0
    n = re.search(r"(\d+)", s)
    return float(n.group(1)) if n else math.nan


def _band(props: dict, layer: str) -> tuple[float, float]:
    """(lo, hi) feet for a feature — each layer codes its limits differently."""
    if layer == "bacc":
        return parse_alt_ft(props.get("lower"), True), parse_alt_ft(
            props.get("upper"), True
        )
    if layer == "subsector":  # no coded band — horizontal only
        return 0.0, math.inf
    if layer == "tma":
        return parse_alt_ft(props.get("lower")), parse_alt_ft(props.get("upper"))
    if layer == "ctr":
        return (
            parse_alt_ft(props.get("lower_1", props.get("lowerlimit"))),
            parse_alt_ft(props.get("upper_1", props.get("upperlimit"))),
        )
    # pdr
    return parse_alt_ft(props.get("lowerlimit")), parse_alt_ft(
        props.get("upperlimit")
    )


def _label(props: dict, layer: str) -> str:
    if layer == "pdr":
        ident = str(props.get("ident") or "").strip()
        name = str(props.get("name") or "").strip()
        return " ".join(p for p in (ident, name) if p) or "PDR"
    return str(props.get("name") or props.get("ident") or "").strip() or (
        layer.upper()
    )


def _title_zone(name: str) -> str:
    """"BANGKOK TMA" -> "Bangkok TMA"; idents with digits (VTD16) kept as-is."""
    out = []
    for w in name.split():
        if any(ch.isdigit() for ch in w):
            out.append(w)
        elif w.upper() in _ZONE_ABBR:
            out.append(w.upper())
        else:
            out.append(w[:1].upper() + w[1:].lower())
    return " ".join(out)


@dataclass
class _Entry:
    label: str
    lo: float
    hi: float


class AirspaceIndex:
    """Per-layer feature metadata + an STRtree over the polygons.

    Whole flights are tagged in one vectorized tree query per layer, then the
    (point, feature) candidate pairs are resolved in file order so the label
    matches the web engine exactly.
    """

    def __init__(
        self,
        entries: dict[str, list[_Entry]],
        trees: dict[str, "shapely.STRtree | None"],
    ):
        self.entries = entries
        self.trees = trees

    def tag(
        self,
        lons: Sequence[float],
        lats: Sequence[float],
        alts_ft: Sequence[float | None],
    ) -> list[str]:
        """Altitude-aware compact sector label for every point of a flight — the
        volume that actually CONTAINS the aircraft at its altitude (a plane at
        FL350 is not in a TMA that tops at 11 000 ft)."""
        n = len(lons)
        pts = shapely.points(
            np.asarray(lons, dtype=float), np.asarray(lats, dtype=float)
        )
        # hits[layer][point] -> labels the point is inside, in file order.
        hits: dict[str, list[list[str]]] = {}
        for layer, _ in _LAYERS:
            per: list[list[str]] = [[] for _ in range(n)]
            tree = self.trees.get(layer)
            entries = self.entries.get(layer, [])
            if tree is not None and n:
                pt_i, ft_i = tree.query(pts, predicate="intersects")
                # Resolve candidates in ascending feature order per point —
                # the web's "first feature wins" rule for non-PDR layers.
                for k in np.lexsort((ft_i, pt_i)):
                    p = int(pt_i[k])
                    e = entries[int(ft_i[k])]
                    alt = alts_ft[p]
                    in_band = (
                        alt is None
                        or layer == "subsector"  # no coded band, horizontal only
                        or (e.lo <= alt <= e.hi)
                    )
                    if in_band and (layer == "pdr" or not per[p]):
                        per[p].append(e.label)
            hits[layer] = per

        def compact(p: int) -> str:
            """The ONE airspace that owns the aircraft here, by _HIERARCHY."""
            for layer in _HIERARCHY:
                got = hits[layer][p]
                if not got:
                    continue
                if layer == "pdr":
                    return ",".join(s.split(" ")[0] for s in got)
                if layer in ("ctr", "tma"):
                    return _title_zone(got[0])
                # 3S/6S are modelled as two altitude slabs (3S_lower/3S_upper);
                # a target is called just "3S"/"6S" (per BACC ops) — the
                # altitude test already picked the right slab. Drop the suffix.
                return re.sub(r"_(lower|upper)$", "", got[0], flags=re.I)
            return ""

        return [compact(p) for p in range(n)]


_INDEX: AirspaceIndex | None = None
_INDEX_FAILED = False


def _load_index() -> AirspaceIndex | None:
    """Build the singleton index; None (and stay quiet) if the data files are
    absent so exports still work in a deployment without the web assets."""
    global _INDEX, _INDEX_FAILED
    if _INDEX is not None or _INDEX_FAILED:
        return _INDEX
    entries: dict[str, list[_Entry]] = {}
    trees: dict[str, shapely.STRtree | None] = {}
    try:
        for layer, fname in _LAYERS:
            path = _SECTORS_DIR / f"{fname}.geojson"
            fc = json.loads(path.read_text(encoding="utf-8"))
            es: list[_Entry] = []
            geoms = []
            for f in fc.get("features", []):
                geom_json = f.get("geometry")
                if not geom_json or geom_json.get("type") not in (
                    "Polygon",
                    "MultiPolygon",
                ):
                    continue
                props = f.get("properties") or {}
                lo, hi = _band(props, layer)
                es.append(
                    _Entry(
                        label=_label(props, layer),
                        lo=lo if not math.isnan(lo) else -math.inf,
                        hi=hi if not math.isnan(hi) else math.inf,
                    )
                )
                geoms.append(shape(geom_json))
            entries[layer] = es
            trees[layer] = shapely.STRtree(geoms) if geoms else None
        _INDEX = AirspaceIndex(entries, trees)
    except (OSError, ValueError, KeyError):
        _INDEX_FAILED = True
        return None
    return _INDEX


def sector_columns(
    lons: Sequence[float],
    lats: Sequence[float],
    alts_ft: Sequence[float | None],
) -> list[str]:
    """Per-point altitude-aware sector label for a whole flight — one compact
    string per input point ("" when the point is outside every volume at its
    altitude, or when the sector data isn't available)."""
    lons = list(lons)
    lats = list(lats)
    alts = list(alts_ft)
    idx = _load_index()
    if idx is None:
        return [""] * len(lons)
    return idx.tag(lons, lats, alts)
